src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs

```diff
@@ -217,9 +217,21 @@ public void ShutdownConnectedNodes(bool enableReuse)
         /// </summary>
         public void ShutdownAllNodes()
         {
-            ShutdownAllNodes(NodeProviderOutOfProc.GetHostHandshake(ComponentHost.BuildParameters.EnableNodeReuse), NodeProviderOutOfProc.GetClientHandshake(), NodeContextTerminated);
+            bool nodeReuse = ComponentHost.BuildParameters.EnableNodeReuse;
+
+            // To avoid issues with mismatched priorities not shutting
+            // down all the nodes on exit, we will attempt to shutdown
+            // all matching notes with and without the priroity bit set.
+            // So precompute both versions of the handshake now.
+            long hostHandshake = NodeProviderOutOfProc.GetHostHandshake(nodeReuse, enableLowPriority: false);
+            long hostHandshakeWithLow = NodeProviderOutOfProc.GetHostHandshake(nodeReuse, enableLowPriority: true);
```

Comment:
I think it makes more sense to only shut down matching nodes (at least if the node has started doing work). If I want to build two projects at once, one with high priority and one with low priority, shutting down one project will inadvertently shut down the other, too. I don't see a reason to want to build two projects of the same priority at once, since they'd just be using each other's resources, but with one high priority and one low priority, you might build a big thing slowly and do some small concept testing in the meantime at higher priority.

Reply:
This API does not affect currently-running builds. It exists so that MSBuild or Visual Studio can shut down _waiting, idle_ nodes left around from recent builds but not currently being used. Since you can't know what priority "all nodes" were created at, you can't pass it in here.

https://github.com/microsoft/msbuild/pull/4162#discussion_r304111430

Reply:
Makes sense to me. Thanks for explaining.